parlai/core/tod/tod_agents.py

```diff
@@ -116,6 +116,438 @@ def _get_agent_type_suffix(self) -> str:
         )
 
 
+######### Agents that dump information from a dataset as gold (explicitly should *not* be used with teachers)
+class _TodDataDumpAgent(TodStructuredDataParser):
+    """
+    For agents which dump data from some dataset, without training/other modifications.
+
+    Implements an "epoch done"
+
+    Member variables assumed to be set in init downstream:
+        self.fold
+    """
+
+    def __init__(self, opt: Opt, shared=None):
+        super().__init__(opt, shared)
+        self.epochDone = False
+        self.batchsize = opt.get("batchsize", 1)
+        self.max_episodes = len(self.episodes)
+        if opt.get("num_episodes", 0) > 0:
+            self.max_episodes = min(self.max_episodes, opt.get("num_episodes"))
+        self.episode_idx = opt.get("batchindex", 0)
+        self._setup_next_episode()
+        self.round_idx = 0  # for some downstream utt + sysUttAndApiCallAgents.
+        if is_distributed():  # cause gotta manually handle
+            rank = get_rank()
+            chunk_size = ceil(self.max_episodes / num_workers())
+            self.episode_idx += rank * chunk_size
+            self.max_episodes = min(self.max_episodes, (rank + 1) * chunk_size)
+
+    def _setup_next_episode(self):
+        self.epochDone = not self.episode_idx < self.max_episodes
+        self.episode = None
+        if not self.epochDone:
+            self.episode = self.episodes[self.episode_idx]
+        self.round_idx = (
+            0  # so downstream agents know which round they are in. Update in `act()`
+        )
+
+    def epoch_done(self) -> bool:
+        return self.epochDone
+
+    def episode_done(self) -> bool:
+        """
+        This is not actually "episode_done" so much as "we want to signify to the world
+        that we have gone past the batch".
+
+        This class should not control whether or not the episode is actually done since
```

Comment:
Ah this is per the conversation structure — It's assumed by convention that the User agent ends the conversation with a [DONE] token... can make this more clear.

Reply:
Edit: After staring at this a bit more, realized it's clearer just editing the code to be more explicit rather than trying to explain what's going on.